libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,646 @@
+      constexpr auto __n = sizeof...(_Views);
+      return __iterator<true>(this, in_place_index<__n - 1>, ranges::end(std::get<__n - 1>(__views_)));
+    } else {
+      return default_sentinel;
+    }
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto size()
+    requires(sized_range<_Views> && ...)
+  {
+    return std::apply(
+        [](auto... __sizes) { return (make_unsigned_t<common_type_t<decltype(__sizes)...>>(__sizes) + ...); },
+        ranges::__tuple_transform(ranges::size, __views_));
+  }
+
+  _LIBCPP_HIDE_FROM_ABI constexpr auto size() const
+    requires(sized_range<const _Views> && ...)
+  {
+    return std::apply(
+        [](auto... __sizes) { return (make_unsigned_t<common_type_t<decltype(__sizes)...>>(__sizes) + ...); },
+        ranges::__tuple_transform(ranges::size, __views_));
+  }
+};
+
+template <class... _Views>
+concat_view(_Views&&...) -> concat_view<views::all_t<_Views>...>;
+
+template <input_range... _Views>
+  requires(view<_Views> && ...) && (sizeof...(_Views) > 0) && __concatable<_Views...>
+template <bool _Const>
+class concat_view<_Views...>::__iterator {
+public:
+  constexpr static bool derive_pack_random_iterator =
+      __derived_from_pack<typename iterator_traits<iterator_t<__maybe_const<_Const, _Views>>>::iterator_category...,
+                          random_access_iterator_tag>;
+  constexpr static bool derive_pack_bidirectional_iterator =
+      __derived_from_pack<typename iterator_traits<iterator_t<__maybe_const<_Const, _Views>>>::iterator_category...,
+                          bidirectional_iterator_tag>;
+  constexpr static bool derive_pack_forward_iterator =
+      __derived_from_pack<typename iterator_traits< iterator_t<__maybe_const<_Const, _Views>>>::iterator_category...,
+                          forward_iterator_tag>;
+  using iterator_category =
+      _If<!is_reference_v<__concat_reference_t<__maybe_const<_Const, _Views>...>>,
+          input_iterator_tag,
+          _If<derive_pack_random_iterator,
+              random_access_iterator_tag,
+              _If<derive_pack_bidirectional_iterator,
+                  bidirectional_iterator_tag,
+                  _If<derive_pack_forward_iterator, forward_iterator_tag, input_iterator_tag > > > >;
+  using iterator_concept =
+      _If<__concat_is_random_access<_Const, _Views...>,
+          random_access_iterator_tag,
+          _If<__concat_is_bidirectional<_Const, _Views...>,
+              bidirectional_iterator_tag,
+              _If< __all_forward<_Const, _Views...>, forward_iterator_tag, input_iterator_tag > > >;
+  using value_type                  = __concat_value_t<__maybe_const<_Const, _Views>...>;
+  using difference_type             = common_type_t<range_difference_t<__maybe_const<_Const, _Views>>...>;
+  using __base_iter _LIBCPP_NODEBUG = variant<iterator_t<__maybe_const<_Const, _Views>>...>;
+
+  __base_iter __it_;
+  __maybe_const<_Const, concat_view>* __parent_ = nullptr;
+
+  template <std::size_t _Idx>
```

Comment:
```suggestion
  template <size_t _Idx>
```

Ditto for other `std::size_t`.